src/qibolab/instruments/icarusqfpga.py

```diff
@@ -1,259 +1,453 @@
+        delay_samples_offset_adc: int = 0,
+        analog_settings: List["dict[str, int]"] = [],
+    ):
+        super().__init__(name, address)
+        self.device = IcarusQRFSoC(address, port)
 
+        self.device.dac_sampling_rate = dac_sampling_rate
+        self.device.adc_sampling_rate = adc_sampling_rate
 
-class PulseBlaster(Instrument):
-    """Driver for the 24-pin PulseBlaster TTL signal generator."""
+        self.channel_delay_offset_dac = delay_samples_offset_dac
+        self.channel_delay_offset_adc = delay_samples_offset_adc
 
-    def __init__(self, name, address, port=5000):
-        super().__init__(name, address)
-        self.port = port
-        self._pins = None
+        for dac in range(self.device.dac_nchannels):
+            self.device.dac[dac].delay = delay_samples_offset_dac
+        for adc in range(self.device.adc_nchannels):
+            self.device.adc[adc].delay = delay_samples_offset_adc
+        self.device.set_adc_trigger_mode(TRIGGER_MODE.SLAVE)
 
-    def setup(self, holdtime_ns, pins=None, **kwargs):
-        """Setup the PulseBlaster.
+        for channel_settings in analog_settings:
+            self.device.set_channel_analog_settings(**channel_settings)
 
-        Arguments:
-            holdtime_ns (int): TTL pulse length and delay between TTL pulses. The experiment repetition frequency is 1 / (2 * holdtime_ns).
-            pins (int): Pins to trigger in hex, defaults to all pins.
-        """
-        if pins is None:
-            pins = list(range(24))
-        self._pins = self._hexify(pins)
-        self._holdtime = holdtime_ns
+        self.device.set_adc_trigger_mode(TRIGGER_MODE.SLAVE)
 
-    def arm(self, nshots, readout_start=0):
-        """Arm the PulseBlaster for playback. Sends a signal to the instrument
-        to setup the pulse sequence and repetition.
+    def setup(self):
+        pass
+
+    def play(
+        self,
+        qubits: Dict[QubitId, Qubit],
+        couplers,
+        sequence: PulseSequence,
+        options: ExecutionParameters,
+    ):
+        """Plays the given pulse sequence without acquisition.
 
         Arguments:
-            nshots (int): Number of TTL triggers to repeat.
+            qubits (dict): Dictionary of qubit IDs mapped to qubit objects.
+            sequence (PulseSequence): Pulse sequence to be played on this instrument.
+            options (ExecutionParameters): Execution parameters for readout and repetition.
         """
-        payload = f"setup,{self._pins},{nshots},{self._holdtime}"
-        return self._send(payload.encode("utf-8"), True)
 
-    def fire(self):
-        """Starts the PulseBlaster."""
-        self._send(b"fire")
-
-    def start(self):
-        pass
+        waveform_array = {dac.id: np.zeros(dac.max_samples) for dac in self.device.dac}
+
+        dac_end_addr = {dac.id: 0 for dac in self.device.dac}
+        dac_sampling_rate = self.device.dac_sampling_rate * 1e6
+        dac_sr_ghz = dac_sampling_rate / 1e9
```

Comment:
> * the `pulse.start` is an integer number of nanoseconds (1e9 in, since it's not seconds)

Following #248, `pulse.start` (and potentially `pulse.duration`) is not an integer in ns for IcarusQ. This is an issue with a `Pulse` library where we "declare" start and duration as `int`s but we don't actually enforce it (luckily Python allows that). 

If we keep going with times (in ns), we should allow non-integer ns in the abstract `Pulse` library and if some driver requires integers to work, do the casting internally in the driver.

Another approach would be to use the number of samples (=duration / sampling rate) as unit of time. This is for sure always integer, however I am not sure how easy it would be in terms of usability.

> However, since we're always going to be at the ns/GHz scale, maybe we could decide to do just the minimum amount of conversions towards GHz, what do you think? 

This sounds reasonable to me, but in the end it's a matter of the driver. Here I see that the sampling rates are hardcoded in MHz
https://github.com/qiboteam/qibolab/blob/b87ea2ce681a94378918aa6d0cccff8079f94cf4/src/qibolab/instruments/icarusqfpga.py#L22
then converted to Hz
https://github.com/qiboteam/qibolab/blob/b87ea2ce681a94378918aa6d0cccff8079f94cf4/src/qibolab/instruments/icarusqfpga.py#L90
and then converted to GHz
https://github.com/qiboteam/qibolab/blob/b87ea2ce681a94378918aa6d0cccff8079f94cf4/src/qibolab/instruments/icarusqfpga.py#L99
to match the units of `pulse.start` in ns. Some steps can probably be simplified. In other drivers we hardcode the sampling rates in GHz directly.

Reply:
Out of your comment, I see two sensible options:
1. forget about integers until the device, and use ns/GHz but just with floats
2. let's go for duration in samples (but the sampling rate will still be float)

I find the second one more confusing, since I foresee comparing two devices and getting a factor wrong just because they have different sampling rates. But if anyone has a strong argument for it, it could be even fine.